### DISCUSSION TITLE
How to replace a Dataset object with m-ld?

### DISCUSSION THREAD

**Discussion by danielbeeke** [m-ld/m-ld-spec#117] [How-to]
I wonder how to swap a Dataset object with m-ld. I have created: https://shacl-form.mediaworks.global/0
I want to make the form collaborative.

It works with a Dataset that conforms to the RDF/JS spec.
I now wonder could I just replace the dataset and would it automatically work? Or do I also need to manually do read and writes on meld?

**Comment by gsvarovsky** [comment]
Hi Daniel, welcome!

A Javascript m-ld 'clone' implements RDF/JS Source, with additional methods for SPARQL queries:
https://js.m-ld.org/interfaces/queryablerdfsource.html. Of course, first you have to create the clone (a local dataset), to get this interface.

At the moment writes have to be done using the native JSON-LD-based format (json-rql). It's JSON-LD with delete/insert keywords. If you'd like to write triples, we could add that pretty easily – in fact it's possible we might want that ourselves, soon.

We're working on some much improved getting-started materials right now. You can preview here: https://m-ld-js-git-getting-started-overhaul-gsvarovsky.vercel.app/. Doesn't talk about RDF yet, but I would be happy to help join the dots.

**Reply by danielbeeke** [reply]
Thanks! Yes triples would be nice as the underlying system seems to be that. At least when I use it in the browser. I saw that QuadStore was underneath it. Yes, like you said in the Gitter chat, the challenge is that DatasetCore is sync.

The nicest way would be that there would be a sync match, but maybe that is just not compatible with the m-ld idea.

Adding triples would be very nice. I am using Grapoi in my form system which also uses DatasetCore.
See https://github.com/OM-MediaWorks/shacl-form for the code. My idea is to make the store injectable. 
I have been playing with a proxy as a store that one additions and deletions does the JSON-LD inserts into m-ld.

I have not got it working 100%, somehow some deletions are missing, probably on my side a mistake. But apart from those bugs, the syncing idea might work too.

**Reply by gsvarovsky** [reply]
OK great! here is what I have in mind for triples: we could broadly augment [MeldState](https://js.m-ld.org/interfaces/meldstate.html) with the [RDF/JS Streams **Store** interface](http://rdf.js.org/stream-spec/#store-interface). That would allow you to asynchronously feed triples into m-ld.

One consideration: in m-ld, an individual call to `write` takes some json-rql representing multiple triple deletions and insertions, making a consistent atomic change according to relevant constraints. The stream-based Store interface doesn't have any such boundaries: you can add and remove individual triples anytime. We could have a transaction for each individual triple, and this might be okay in some cases, but in others it might represent an intermediate state between consistent ones. So I'd prefer to think up front about having some kind of coherent 'write' boundary.

One way of doing this would be to require the use of SPARQL Update – so the coherent write is a SPARQL statement; which could be expressed as a string, or using the algebra.

Another way would be to expose the `Store` interface only to a callback. So, for example, a 'write' might look like this:

```js
await state.writeRdf(async store => { /* make calls to the store here */ });
```

**Reply by danielbeeke** [reply]
Thanks for the great descriptions! I like the last idea, the store via callback. I think it gives a developer super powers to be able to add triples.

**Reply by gsvarovsky** [reply]
OK nice, I'll experiment with the Store-callback API 👍

**Reply by gsvarovsky** [reply]
We will also need to think about what to do with some other features, for which the data representation in the json-rql API is abstracted away from RDF, and would need more thought. These are Lists, Binary Data and Collaborative Text. I think your form had examples of all three! For now perhaps let's just find out which ones you want to use.

Lists in m-ld, using the `@list` JSON-LD keyword, do not transform to RDF Collections. They are more like RDF Containers (but still, not quite 🙂). Plain RDF Collections don't converge nicely with concurrent edits – more about this in our [Semantics poster](https://bit.ly/realtime-rdf-poster). If you want to use lists we'll have to consider what to do here.

Binary data is not a big deal because it's only an efficiency gain, you can always use xs:base64Binary, and we can probably easily create a custom data factory to even take away the need for the API client to encode anything. Or you may want to handle attachments completely separately anyway.

Collaborative text will probably be useful for description fields, so multiple users can edit the field at the same time. If we want to use this feature it's not going to be pure RDF any more.

**Reply by gsvarovsky** [reply]
- https://github.com/m-ld/m-ld-js/issues/81

**Reply by danielbeeke** [reply]
> Lists in m-ld, using the @list JSON-LD keyword, do not transform to RDF Collections. They are more like RDF Containers (but still, not quite 🙂). Plain RDF Collections don't converge nicely with concurrent edits – more about this in our [Semantics poster](https://bit.ly/realtime-rdf-poster). If you want to use lists we'll have to consider what to do here.

I read the semantics poster. 
I can adapt the system to also support the structure m-ld is using. Another idea could be to abstract lists away. Do I see it correct that I could propagate list changes in m-ld specific format, but keep them in my own dataset as a rdf list?

Yes attachments should be IRIs and then on the storage end a cron task should run to clean up unused attachments.

Super cool stuff :)

**Reply by gsvarovsky** [reply]
- https://github.com/m-ld/m-ld-js/issues/172
